### PR TITLE
Optimize how we calculate `likely_domains` during backfill

### PR DESCRIPTION
**Pull request by MadLittleMods**
Optimize how we calculate `likely_domains` during backfill because I've seen this take 17s in production just to `get_current_state` which is used to `get_domains_from_state` (see case [*2. Loading tons of events* in the `/messages` investigation issue](https://github.com/matrix-org/synapse/issues/13356)).

There are 3 ways we currently calculate hosts that are in the room:

 1. `get_current_state` -> `get_domains_from_state`
    - Used in `backfill` to calculate `likely_domains` and `/timestamp_to_event` because it was cargo-culted from `backfill`
    - This one is being eliminated in favor of `get_current_hosts_in_room` in this PR 🕳
 1. `get_current_hosts_in_room`
    - Used for other federation things like sending read receipts and typing indicators
 1. `get_hosts_in_room_at_events`
    - Used when pushing out events over federation to other servers in the `_process_event_queue_loop`

Fix https://github.com/matrix-org/synapse/issues/13626

Part of https://github.com/matrix-org/synapse/issues/13356

Mentioned in [internal doc](https://docs.google.com/document/d/1lvUoVfYUiy6UaHB6Rb4HicjaJAU40-APue9Q4vzuW3c/edit#bookmark=id.2tvwz3yhcafh)

#### Before

The query from `get_current_state` sucks just because we have to get all 80k events. And we see almost the exact same performance locally trying to get all of these events (16s vs 17s):
```
synapse=# SELECT type, state_key, event_id FROM current_state_events WHERE room_id = '!OGEhHVWSdvArJzumhm:matrix.org';
Time: 16035.612 ms (00:16.036)

synapse=# SELECT type, state_key, event_id FROM current_state_events WHERE room_id = '!OGEhHVWSdvArJzumhm:matrix.org';
Time: 4243.237 ms (00:04.243)
```

But what about `get_current_hosts_in_room`: When there is 8M rows in the `current_state_events` table, the previous query in `get_current_hosts_in_room` took 13s from complete freshness (when the events were first added). But takes 930ms after a Postgres restart or 390ms if running back to back to back.

```sh
$ psql synapse
synapse=# \timing on
synapse=# SELECT COUNT(DISTINCT substring(state_key FROM '@[^:]*:(.*)$'))
FROM current_state_events
WHERE
    type = 'm.room.member'
    AND membership = 'join'
    AND room_id = '!OGEhHVWSdvArJzumhm:matrix.org';
 count
-------
  4130
(1 row)

Time: 13181.598 ms (00:13.182)

synapse=# SELECT COUNT(*) from current_state_events where room_id = '!OGEhHVWSdvArJzumhm:matrix.org';
 count
-------
 80814

synapse=# SELECT COUNT(*) from current_state_events;
  count
---------
 8162847

synapse=# SELECT pg_size_pretty( pg_total_relation_size('current_state_events') );
 pg_size_pretty
----------------
 4702 MB
```

#### After

I'm not sure how long it takes from complete freshness as I only really get that opportunity once (maybe restarting computer but that's cumbersome) and it's not really relevant to normal operating times. Maybe you get closer to the fresh times the more access variability there is so that Postgres caches aren't as exact. Update: The longest I've seen this run for is 6.4s and 4.5s after a computer restart.

After a Postgres restart, it takes 330ms and running back to back takes 260ms.

```sh
$ psql synapse
synapse=# \timing on
Timing is on.
synapse=# SELECT
    substring(c.state_key FROM '@[^:]*:(.*)$') as host
FROM current_state_events c
/* Get the depth of the event from the events table */
INNER JOIN events AS e USING (event_id)
WHERE
    c.type = 'm.room.member'
    AND c.membership = 'join'
    AND c.room_id = '!OGEhHVWSdvArJzumhm:matrix.org'
GROUP BY host
ORDER BY min(e.depth) ASC;
Time: 333.800 ms
```

#### Going further

To improve things further we could add a `limit` parameter to `get_current_hosts_in_room`. Realistically, we don't need 4k domains to choose from because there is no way we're going to query that many before we a) probably get an answer or b) we give up. 

Another thing we can do is optimize the query to use a index skip scan:

 - https://wiki.postgresql.org/wiki/Loose_indexscan
 - Index Skip Scan, https://commitfest.postgresql.org/37/1741/
 - https://www.timescale.com/blog/how-we-made-distinct-queries-up-to-8000x-faster-on-postgresql/

### Dev notes

Created a utility script to dump the state events from a 
`GET https://matrix-client.matrix.org/_matrix/client/r0/rooms/!OGEhHVWSdvArJzumhm:matrix.org/state` response for testing in your local dev database, https://github.com/MadLittleMods/matrix-synapse-state-dumper

It can also add in filler events in between in the database so it isn't such an easy sequential read for Postgres.

It's rough and just barely does what I needed to get some testing for this PR. Not very general.

---

```
brew services restart postgresql@13

$ psql synapse
> \timing on
> ...
> \q
```

---

Get count of rows when using `group by`:
```
SELECT COUNT(*) OVER () AS TotalRecords
FROM current_state_events c
/* Get the depth of the event from the events table */
INNER JOIN events AS e USING (event_id)
WHERE
    c.type = 'm.room.member'
    AND c.membership = 'join'
    AND c.room_id = '!OGEhHVWSdvArJzumhm:matrix.org'
GROUP BY substring(c.state_key FROM '@[^:]*:(.*)$')
ORDER BY min(e.depth);
```

---

<details>
<summary><code>BackfillBenchmarkTestCase</code> <code>test_benchmark_likely_domains</code></summary>

```py
class BackfillBenchmarkTestCase(unittest.HomeserverTestCase):
    def prepare(self, reactor, clock, hs):
        self._storage_controllers = hs.get_storage_controllers()

    def test_benchmark_likely_domains(self):
        import time
        import logging

        logger = logging.getLogger(__name__)

        from synapse.handlers.federation import get_domains_from_state

        async def calculate_likely_domains():
            # Find other homeservers from the given state in the room
            curr_state = await self._storage_controllers.state.get_current_state(
                "!OGEhHVWSdvArJzumhm:matrix.org"
            )
            curr_domains = get_domains_from_state(curr_state)
            likely_domains = [
                domain for domain, depth in curr_domains if domain != self.server_name
            ]
            return likely_domains

        def time_calculate_likely_domains(test_prefix: str):
            benchmark_start_time = time.time()
            likely_domains = self.get_success(calculate_likely_domains())
            benchmark_end_time = time.time()
            logger.info(
                "Benchmark time (%s): %s",
                "{test_prefix: <13}".format(test_prefix=test_prefix),
                (benchmark_end_time - benchmark_start_time),
            )
            self.assertGreater(len(likely_domains), 4000)

        time_calculate_likely_domains("1")
        time_calculate_likely_domains("2")
        time_calculate_likely_domains("3")
```

</details>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/handlers/federation.py
Comment (MadLittleMods) [inline synapse/handlers/federation.py:102]: Eliminating this option to get domains for a room in favor of `get_current_hosts_in_room`.

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [inline synapse/storage/databases/main/roommember.py:1059]: Updating the query to order by depth to match what `get_domains_from_state` did before (drop-in replacement).

The heuristic of sorting by servers who have been there the longest is good because they're most likely to have anything we ask about.

File: synapse/handlers/room.py
Comment (MadLittleMods) [inline synapse/handlers/room.py:1466]: I think these CI failures are because we're just querying our own server when we want some other federated one. `get_current_hosts_in_room` doesn't exclude our own server.

 - https://github.com/matrix-org/synapse/runs/7929900798?check_suite_focus=true#step:4:3822
 - https://github.com/matrix-org/synapse/runs/7929900765?check_suite_focus=true#step:4:3780

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [inline synapse/storage/databases/main/roommember.py:1035]: I guess this isn't a total drop-in replacement as before since we don't do any sorting for this short-cut grabbing the users from cache. Can we have any assumption on how the `get_users_in_room.cache` is ordered?

Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:1035]: Updated `get_users_in_room` to return users sorted by lowest -> highest depth. And the `set` comprehension works as expected so duplicates don't change the order.

```py
users = ["foo", "bar", "baz", "foo"]
domains = {u for u in users}

print(f"domains={domains}")
#  -> domains={'foo', 'bar', 'baz'}
```

File: synapse/storage/databases/main/roommember.py
Comment (reivilibre) [inline synapse/storage/databases/main/roommember.py:1062]: This is not true; sets aren't ordered collections and their elements don't have any prescribed order.

(Python gives a misleading impression because CPython's *dicts* are ordered based on insertion order, but that's because of a CPython implementation detail that it uses a set of entries alongside a linked list)

As an example:
```
>>> list(set([2, 1]))
[1, 2]
```

You'll have to deduplicate some other way if the ordering is important

**Review by reivilibre — CHANGES_REQUESTED** [review]
Interesting optimisation!

Would be worth checking how much we're going to negatively impact callers that don't care about the depth ordering ­— introducing a join (not even with all the columns covered by an index) in a section of the code that says 'Avoiding a join is a Very Good Thing' makes me worry a bit :).

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:215]: Time-wise, I think by the nature of the extra complexity, it will take more time but it's not something I optimized for here. If we don't like this trade-off, we could remove the cache-shortcut lookup in `get_current_hosts_in_room` around it.

Is there another concern about the order itself?

File: synapse/storage/databases/main/roommember.py
Comment (DMRobertson) [reply synapse/storage/databases/main/roommember.py:1062]: Perhaps https://grantjenks.com/docs/sortedcollections/orderedset.html ? We already use SortedDict and SortedList elsewhere in Synapse

Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:1062]: ohhh, good shout! It looks like I also got lucky in my [other example](https://github.com/matrix-org/synapse/pull/13575/files#r952081093). Running it a few more times and I can see the ordering change around 😵

I guess we will have to return a `List[str]` here for the sort guarantee for this function in general

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:1056]: The `set` comprehension actually doesn't maintain the order. Just got lucky when trying this out the first time, see https://github.com/matrix-org/synapse/pull/13575#discussion_r952748930

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:215]: fwiw, here are some first-run timings. Not sure how much the first one affects the second one 🤷 

```
$ psql synapse
synapse=# \timing on

synapse=# SELECT c.state_key FROM current_state_events as c
/* Get the depth of the event from the events table */
INNER JOIN events AS e USING (event_id)
WHERE c.type = 'm.room.member' AND c.room_id = '!OGEhHVWSdvArJzumhm:matrix.org' AND membership = 'join'
/* Sorted by lowest depth first */
ORDER BY e.depth ASC;

Time: 4374.227 ms (00:04.374)


synapse=# SELECT c.state_key FROM current_state_events as c
WHERE c.type = 'm.room.member' AND c.room_id = '!OGEhHVWSdvArJzumhm:matrix.org' AND membership = 'join';

Time: 1947.409 ms (00:01.947)
```

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py]: Updated to use deduplicated `List[str]` 👍

Can switch to an ordered set if we decide to use that ⏩

File: synapse/storage/databases/main/roommember.py
Comment (reivilibre) [inline synapse/storage/databases/main/roommember.py:1072]: This will be O(n^2) slow because `domain not in domains` is an expensive `O(n)` thing to do.
A technique you might want to use is to maintain a set of the domains you've already seen...

```suggestion
            domains: List[str] = []
            processed_domains: Set[str] = set()
            for u in users:
                domain = get_domain_from_id(u)
                if domain not in processed_domains:
                    domains.append(domain)
                    processed_domains.add(domain)
```

File: synapse/storage/databases/main/roommember.py
Comment (reivilibre) [reply synapse/storage/databases/main/roommember.py:215]: Only doubling isn't as bad as I was expecting, so maybe it's fine. Possibly worth checking a graph for how much time is spent on `get_users_in_rooms` and thinking if it would burn us if that time were doubled.

File: synapse/storage/databases/main/roommember.py
Comment (reivilibre) [reply synapse/storage/databases/main/roommember.py]: as an aside, confusing as that grantjenks has both 'sorted collections' and 'sorted containers' and they have different data structures...

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:1072]: Any preference on having a separate `Set` for fast lookups vs using a `Dict` which you mentioned maintains insertion order and just call `dict.values()` at the end?

Comment (reivilibre) [reply synapse/storage/databases/main/roommember.py]: I take it you mean `.keys()`, but I think `Dict` maintaining insertion order is a CPython quirk and not guaranteed by the Python standard, so I'd rather not rely on it. We do have some admins running unusual interpreters :-)

Comment (DMRobertson) [reply synapse/storage/databases/main/roommember.py]: > `Dict` maintaining insertion order is a CPython quirk and not guaranteed by the Python standard

It's guaranteed from 3.7 upwards. See "Python data model improvements" in https://docs.python.org/3/whatsnew/3.7.html#summary-release-highlights

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:215]: I've looked at the usages and there are some areas where it doesn't feel great to increase the time (calculating if an appservice is interested in event (this could probably be smarter though anyway), calculating `join_authorised_via_users_server` in a `/make_join` (idk how important this is), newly joined rooms in `/sync`). Overall seems kinda fine ⏩

I can create separate PRs to fix up the following areas first, then we can merge this after those places are changed.

There are a few areas where it feels like we shouldn't be using `get_users_in_room` and can just use a direct lookup for the single user:

 - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/events.py#L176-L177
 - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/room.py#L1287-L1288
  - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/message.py#L764-L765
 - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/room_member.py#L1623-L1624 
 - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/server_notices/server_notices_manager.py#L115

 There are a few places where we should use `get_current_hosts_in_room`:
 
  - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/directory.py#L86-L87
  - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/directory.py#L290-L291
  - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/presence.py#L2054-L2055
  - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/typing.py#L365-L366

Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:215]: Addressed these mis-uses in:

 - https://github.com/matrix-org/synapse/pull/13605
 - https://github.com/matrix-org/synapse/pull/13608

Comment (reivilibre) [reply synapse/storage/databases/main/roommember.py:215]: I'd expect `get_users_in_room` to be visible in the Database section of the Grafana dashboard, but it seems like it might be so insignificant that it doesn't show up there. I think that probably hints at the answer.

File: synapse/storage/databases/main/roommember.py
Comment (reivilibre) [reply synapse/storage/databases/main/roommember.py]: > It's guaranteed from 3.7 upwards (which is the oldest Python we support). See "Python data model improvements" in [docs.python.org/3/whatsnew/3.7.html#summary-release-highlights](https://docs.python.org/3/whatsnew/3.7.html#summary-release-highlights)

good to know. In that case, feel free to do whichever is best since 3.7 is our minimum version.

**Review by reivilibre — APPROVED** [review]
LGTM — suppose it would be prudent to land this after fixing up the (mis-)usages that you identified

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:215]: Seems mostly at `0.5%` of total time, sometimes `2%`. But it's down in the noise of the rest of the txn's we do if you show all of them again.

Doubling 3ms - 15ms average time isn't horrible but would be good to check in on this after this PR merges.

File: changelog.d/13575.misc
Comment (MadLittleMods) [inline changelog.d/13575.misc:1]: Waiting to merge this until we get a little more time for our baseline metrics on `matrix.org`

**Comment by MadLittleMods** [comment]
Thanks for the review @reivilibre 🦒

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:215]: Here is some follow-up. I think this PR was rolled out on `matrix.org` (or maybe just the RC was which wouldn't include this?)

The average looks about the same. It seems a bit more peakier on total txn time.

Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:215]: Another follow-up, this time I think the changes are actually deployed to `matrix.org` as indicated by that change in the `UP` graph at the top (and `/versions` shows a hash that includes it)

It looks like it does take significantly more time on average (~100ms vs ~10ms before) but we're doing the query a whole lot more less probably because we fixed up those [mis](https://github.com/matrix-org/synapse/pull/13605)-[uses](https://github.com/matrix-org/synapse/pull/13608).

https://grafana.matrix.org/d/000000012/synapse?orgId=1&from=1662362659334&to=1662535459334&viewPanel=180

**Comment by DMRobertson** [comment]
```
$ git log --oneline -S get_users_in_room v1.66.0..v1.68.0
  e124b24ad 1.68.0rc1
  906cead9c Update docstrings to explain the impact of partial state (#13750)
  cdf7fb737 1.67.0rc1
  1a209efdb Update `get_users_in_room` mis-use to get hosts with dedicated `get_current_hosts_in_room` (#13605)
  d58615c82 Directly lookup local membership instead of getting all members in a room first (`get_users_in_room` mis-use) (#13608)
  c807b814a Use dedicated `get_local_users_in_room` to find local users when calculating `join_authorised_via_users_server` of a `/make_jo  in` request (#13606)
```

On the affected hosts we saw that postgres used a heap scan rather than an index scan for the aggressive query. Forcing an index-only scan (SET enable_bitmapscan to off) made the query quicker (500->300ms) for one trial set of parameters. Unclear if those are representitive. Also unclear if the index-only scan used more or less file IO.

But this might be barking up the wrong tree. Are we perhaps calling `get_users_in_room` in more circumstances now? Do we need to up its cache?

File: synapse/storage/databases/main/roommember.py
Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:215]: This probably has a bigger penalty for other usage patterns, see https://github.com/matrix-org/synapse/issues/13942

Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:215]: Probably fixed https://github.com/matrix-org/synapse/issues/13942 via https://github.com/matrix-org/synapse/pull/13958 (another `get_users_in_room` mis-use)

Fixed up another mis-use in https://github.com/matrix-org/synapse/pull/13960

And found another mis-use that's tracked by https://github.com/matrix-org/synapse/issues/13967. Since this involves more than just switching it out, I'm not going to tackle. Feel free to pick it up!

Comment (MadLittleMods) [reply synapse/storage/databases/main/roommember.py:215]: This ended up as two functions `get_current_hosts_in_room` and `get_current_hosts_in_room_ordered` so that we don't have to take the performance hit of the extra join in `get_users_in_room`, see https://github.com/matrix-org/synapse/pull/13972

**Pull request merged**
